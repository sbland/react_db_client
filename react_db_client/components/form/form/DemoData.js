import { filterTypes } from '@react_db_client/constants.client-types';

export const demoHeadingsData = [
  {
    uid: 'uid',
    label: 'UID',
    type: filterTypes.uid,
  },
  {
    uid: 'name',
    label: 'Text',
    type: filterTypes.text,
    required: true,
  },
  {
    uid: 'longText',
    label: 'Long Text',
    type: filterTypes.textLong,
  },
  {
    uid: 'date',
    label: 'Date',
    type: filterTypes.date,
  },
  {
    uid: 'number',
    label: 'Number',
    type: filterTypes.number,
  },
  {
    uid: 'number-capped',
    label: 'Number Capped',
    type: filterTypes.number,
    min: 3,
    max: 8,
    step: 0.1,
    defaultValue: 5,
  },
  {
    uid: 'demoField',
    label: 'Custom Field Types',
    type: 'demoFieldType',
  },
  {
    uid: 'reference',
    label: 'Reference',
    collection: 'democollection',
    type: filterTypes.reference,
    objectLink: true,
  },
  {
    uid: 'file',
    label: 'File',
    collectionId: 'democollection',
    documentId: 'docid',
    fileType: 'document',
    type: filterTypes.file,
  },
  {
    uid: 'image',
    label: 'Image',
    type: filterTypes.image,
  },
  {
    uid: 'bool',
    label: 'Bool',
    type: filterTypes.bool,
  },
  {
    uid: 'select',
    label: 'Select',
    type: filterTypes.select,
    options: [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'selectreadonly',
    label: 'Select Read Only',
    type: filterTypes.select,
    readOnly: true,
    options: [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'selectSearch',
    label: 'Select Search',
    type: filterTypes.selectSearch,
    searchFieldTargetField: 'label',
    searchFn: async () => [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'selectSearchMulti',
    label: 'Select Search Multi',
    type: filterTypes.selectSearch,
    multiple: true,
    searchFieldTargetField: 'label',
    allowEmptySearch: true,
    searchFn: async () => [
      { uid: 'rep1', label: 'Rep 1' },
      { uid: 'rep2', label: 'Rep 2' },
    ],
  },
  {
    uid: 'multiSelect',
    label: 'Multi Select',
    type: filterTypes.selectMulti,
    options: [
      { uid: 'foo', label: 'Rep 1' },
      { uid: 'bar', label: 'Rep 2' },
      { uid: 'dee', label: 'Rep 3' },
    ],
  },
  {
    uid: 'multiSelectlist',
    label: 'Multi Select No Dropdown',
    type: filterTypes.selectMulti,
    asDropdown: false,
    options: [
      { uid: 'foo', label: 'Rep 1' },
      { uid: 'bar', label: 'Rep 2' },
      { uid: 'dee', label: 'Rep 3' },
    ],
  },
  {
    uid: 'multiSelectlistShowAll',
    label: 'Multi Select Show All',
    type: filterTypes.selectMulti,
    selectType: 'showall',
    options: [
      { uid: 'foo', label: 'Rep 1' },
      { uid: 'bar', label: 'Rep 2' },
      { uid: 'dee', label: 'Rep 3' },
    ],
  },
  {
    uid: 'embedded',
    label: 'Embedded',
    type: filterTypes.embedded,
    children: [{ uid: 'embeddedtext', label: 'Embedded Text', type: filterTypes.text }],
  },
  {
    uid: 'embeddedb',
    label: 'Embedded B',
    type: filterTypes.embedded,
    orientation: 'horiz',
    children: [
      { uid: 'embeddedtog1', label: 'Embedded Tog1', type: filterTypes.bool },
      { uid: 'embeddedtog2', label: 'Embedded Tog2', type: filterTypes.bool },
      { uid: 'embeddedtog3', label: 'Embedded Tog3', type: filterTypes.bool },
    ],
  },
  {
    uid: 'dict',
    label: 'Dictionary Field',
    type: filterTypes.dict,
  },
  {
    uid: 'video',
    label: 'Video     Field',
    type: filterTypes.video,
  }
];

export const demoFormData = {
  name: 'Name 1',
  uid: 'name-1',
  date: '2019-11-02T12:04:44.626+00:00',
  selectreadonly: 'rep1',
  multiSelect: ['foo', 'bar'],
  longText: `Long Text spanning multiple lines
  1
  2
  3
  4
  5
  6
  7
  8
  9
  `,
};

export const demoAdditionalData = {
  customFieldStyle: { background: 'grey', padding: '2px' },
};
