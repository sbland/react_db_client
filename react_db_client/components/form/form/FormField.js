import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { filterTypes } from '@react_db_client/constants.client-types';
import { switchF } from '@react_db_client/helpers.func-tools';

import { FieldReadOnly } from '@samnbuk/react_db_client.components.form.form-fields.field-read-only';
import { FieldLabel } from './field-label';

import './_form.scss';

const defaultComponent = () => FieldReadOnly;

export const FormField = (propsIn) => {
  const { heading, value, updateFormData, additionalData, componentMap } = propsIn;
  const props = useMemo(
    () => ({
      updateFormData,
      value,
      key: `${heading.uid}-sub`,
      additionalData,
      ...heading,
    }),
    [heading, updateFormData, value, additionalData]
  );
  const { label, required, type, uid, hasChanged, readOnly } = heading;

  const FormComponent = useMemo(
    () => switchF(heading.type, componentMap, defaultComponent),
    [heading.type, componentMap, defaultComponent]
  );

  const labelClassName = [
    'form_label',
    `${required ? 'required' : ''}`,
    `${hasChanged ? 'hasChanged' : ''}`,
  ]
    .filter((f) => f)
    .join(' ');
  const rowClassname = ['form_row', `form_row_heading_id_${uid}`].filter((f) => f).join(' ');

  const showLabel = [filterTypes.bool, filterTypes.toggle].indexOf(type) === -1 || heading.readOnly; // we do not need a label for a toggle box
  return (
    <div className={rowClassname} key={uid}>
      <FieldLabel
        uid={uid}
        label={label}
        inputClassName={labelClassName}
        hasChanged={hasChanged}
        required={required}
        hasLabel={showLabel}
      />
      <FormComponent {...props} />
    </div>
  );
};

FormField.propTypes = {
  heading: PropTypes.shape({
    uid: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    unit: PropTypes.string,
    required: PropTypes.bool,
    hasChanged: PropTypes.bool,
    readOnly: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf(['true', 'false'])]),
  }).isRequired,
  value: PropTypes.any,
  updateFormData: PropTypes.func.isRequired,
  additionalData: PropTypes.shape({}).isRequired,
  componentMap: PropTypes.objectOf(PropTypes.elementType),
};

FormField.defaultProps = {
  value: null,
  componentMap: {},
};
