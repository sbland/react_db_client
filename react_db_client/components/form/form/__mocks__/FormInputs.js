import React from 'react';
import PropTypes from 'prop-types';


const FormInputs = ({
  headings,
  formData,
  updateFormData,
  orientation,
  heading,
  isSection,
}) => {
  const className = `
    form_inputs
    ${(isSection) ? 'formSection' : ''}
    ${orientation}
  `;
  return (
    <section className={className}>
      {(!isSection) && <p>* is required.  (!) has been Overriden.</p>}
      <h4 className="formSection_heading">{heading}</h4>
      <div className="form_row">
        <input type="text" />
      </div>
    </section>
  );
};

FormInputs.propTypes = {
  headings: PropTypes.arrayOf(PropTypes.shape({
    uid: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    type: PropTypes.oneOf([
      'uid',
      'reference',
      'file',
      'fileMultiple',
      'textLong',
      'text',
      'number',
      'date',
      'image',
      'bool',
      'select',
      'selectMulti',
      'embedded',
    ]).isRequired,
    unit: PropTypes.string,
  })).isRequired,
  formData: PropTypes.object.isRequired,
  updateFormData: PropTypes.func.isRequired,
  orientation: PropTypes.oneOf(['horiz', 'vert']),
  heading: PropTypes.string,
  isSection: PropTypes.bool,
};

FormInputs.defaultProps = {
  orientation: 'vert',
  heading: '',
  isSection: false,
};

export default FormInputs;
