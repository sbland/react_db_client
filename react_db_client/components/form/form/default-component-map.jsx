import React from 'react';
import { filterTypes } from '@samnbuk/react_db_client.constants.client-types';

import { FieldText } from '@samnbuk/react_db_client.components.form.form-fields.field-text';
import { FieldTextArea } from '@samnbuk/react_db_client.components.form.form-fields.field-text-area';
import { FieldBool } from '@samnbuk/react_db_client.components.form.form-fields.field-bool';
import { FieldDate } from '@samnbuk/react_db_client.components.form.form-fields.field-date';
import { FieldNumber } from '@samnbuk/react_db_client.components.form.form-fields.field-number';
import { FieldObjectRef } from '@samnbuk/react_db_client.components.form.form-fields.field-object-ref';
import { FieldMultiSelect } from '@samnbuk/react_db_client.components.form.form-fields.field-multi-select';
import { FieldSelect } from '@samnbuk/react_db_client.components.form.form-fields.field-select';
import { FieldSelectSearch } from '@samnbuk/react_db_client.components.form.form-fields.field-select-search';
import { FieldFile } from '@samnbuk/react_db_client.components.form.form-fields.field-file';

export const defaultComponentMap = ({ asyncGetDocuments, fileServerUrl } = {}) => ({
  [filterTypes.text]: () => FieldText,
  [filterTypes.select]: () => FieldSelect,
  [filterTypes.selectMulti]: () => FieldMultiSelect,
  // [filterTypes.fileMultiple]: () => FieldFile,
  [filterTypes.file]: () => (props) =>
    <FieldFile {...props} fileServerUrl={fileServerUrl} asyncGetDocuments={asyncGetDocuments} />,
  // [filterTypes.image]: () => FieldFile,
  [filterTypes.textLong]: () => FieldTextArea,
  [filterTypes.number]: () => FieldNumber,
  [filterTypes.date]: () => FieldDate,
  [filterTypes.bool]: () => FieldBool,
  [filterTypes.toggle]: () => FieldBool,
  [filterTypes.selectSearch]: () => FieldSelectSearch,
  [filterTypes.reference]: () => (props) =>
    <FieldObjectRef {...props} asyncGetDocuments={asyncGetDocuments} />,
});
