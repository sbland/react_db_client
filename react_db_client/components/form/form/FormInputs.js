import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { filterTypes } from '@react_db_client/constants.client-types';

import '@samnbuk/react_db_client.constants.style';
import { FormField as DefaultFormField } from './FormField';
import './_form.scss';

export const FormInputs = ({
  FormField,
  headings,
  formData,
  updateFormData,
  orientation,
  heading: sectionHeading,
  isSection,
  showKey,
  additionalData,
  componentMap,
  id,
}) => {
  const className = [
    'form_inputs',
    'formSection',
    `${orientation}`,
    sectionHeading ? 'hasHeading' : '',
    `formSection_${id}`,
  ]
    .filter((f) => f)
    .join(' ');

  const fields = useMemo(
    () =>
      headings.map((heading) => {
        const value =
          formData && formData[heading.uid] !== null
            ? formData[heading.uid]
            : heading.defaultValue || heading.def;

        if (heading.type === filterTypes.embedded) {
          return (
            <FormInputs
              key={heading.uid}
              headings={heading.children}
              formData={formData}
              updateFormData={updateFormData}
              orientation={heading.orientation}
              heading={heading.label}
              additionalData={additionalData}
              componentMap={componentMap}
              FormField={FormField}
              id={heading.uid}
              isSection
            />
          );
        }
        // return <div>{heading.uid}</div>;
        return (
          <FormField
            heading={heading}
            updateFormData={updateFormData}
            value={value}
            key={heading.uid}
            additionalData={additionalData}
            componentMap={componentMap}
          />
        );
      }),
    [headings, formData, updateFormData, additionalData, componentMap]
  );

  return (
    <section className={className}>
      {!isSection && showKey && <p>* is required. (!) has been modified.</p>}
      <h4 className="formSection_heading">{sectionHeading}</h4>
      {fields}
    </section>
  );
};

FormInputs.propTypes = {
  FormField: PropTypes.func,
  headings: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      orientation: PropTypes.oneOf(['horiz', 'vert']),
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      unit: PropTypes.string,
      children: PropTypes.arrayOf(PropTypes.shape({})),
    })
  ).isRequired,
  formData: PropTypes.object.isRequired,
  updateFormData: PropTypes.func.isRequired,
  orientation: PropTypes.oneOf(['horiz', 'vert']),
  heading: PropTypes.string,
  isSection: PropTypes.bool,
  showKey: PropTypes.bool,
  additionalData: PropTypes.shape({}),
  componentMap: PropTypes.objectOf(PropTypes.elementType).isRequired,
  id: PropTypes.any,
};

FormInputs.defaultProps = {
  FormField: DefaultFormField,
  orientation: 'vert',
  heading: '',
  isSection: false,
  showKey: true,
  additionalData: {},
  id: null,
};
