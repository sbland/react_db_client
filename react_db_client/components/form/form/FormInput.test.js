import '@samnbuk/react_db_client.testing.enzyme-setup';
// test('should work', () => {
//     //
// })

import React from 'react';
import { mount, shallow } from 'enzyme';
import { MockReactC } from '@samnbuk/react_db_client.testing.utils';

import { FormInputs } from './FormInputs';
import { demoHeadingsData, demoFormData } from './DemoData';
import { FormField } from './FormField';
import { defaultComponentMap } from './default-component-map';

jest.mock('./FormField', () => MockReactC('FormField', ['FormField']));

const updateFormData = jest.fn();

const defaultProps = {
  headings: demoHeadingsData,
  formData: demoFormData,
  updateFormData,
  additionalData: {},
  componentMap: defaultComponentMap(),
  FormField,
};

describe('FormInputs', () => {
  test('Renders', () => {
    shallow(<FormInputs {...defaultProps} />);
  });
  describe('shallow renders', () => {
    test('Matches Snapshot', () => {
      const component = shallow(<FormInputs {...defaultProps} />);
      const tree = component.debug();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('Unit tests', () => {
    let component;
    beforeEach(() => {
      updateFormData.mockClear();
      component = mount(<FormInputs {...defaultProps} />);
    });
    test('should call updateformData when an input changes', () => {
      const input = component.find(FormField).first();
      const field = 'demofield';
      const value = 'demovalue';
      input.props().updateFormData(field, value);
      expect(updateFormData).toHaveBeenCalledWith(field, value);
    });
    test('should create the correct number of fields', () => {
      const fields = component.find(FormField);
      // Add 2 to length as we have embedded headings
      expect(fields.length).toEqual(demoHeadingsData.length + 2);
    });
  });
});
