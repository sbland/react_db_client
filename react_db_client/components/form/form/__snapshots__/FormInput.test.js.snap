// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInputs shallow renders Matches Snapshot 1`] = `
"<section className=\\"form_inputs formSection vert formSection_null\\">
  <p>
    * is required. (!) has been modified.
  </p>
  <h4 className=\\"formSection_heading\\" />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value=\\"name-1\\" additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value=\\"Name 1\\" additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value=\\"Long Text spanning multiple lines\\\\n  1\\\\n  2\\\\n  3\\\\n  4\\\\n  5\\\\n  6\\\\n  7\\\\n  8\\\\n  9\\\\n  \\" additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value=\\"2019-11-02T12:04:44.626+00:00\\" additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value=\\"rep1\\" additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={{...}} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormField heading={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} value={[undefined]} additionalData={{...}} componentMap={{...}} />
  <FormInputs headings={{...}} formData={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} orientation=\\"vert\\" heading=\\"Embedded\\" additionalData={{...}} componentMap={{...}} FormField={[Function: MockedComponentNamed] { displayName: 'FormField' }} id=\\"embedded\\" isSection={true} showKey={true} />
  <FormInputs headings={{...}} formData={{...}} updateFormData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} orientation=\\"horiz\\" heading=\\"Embedded B\\" additionalData={{...}} componentMap={{...}} FormField={[Function: MockedComponentNamed] { displayName: 'FormField' }} id=\\"embeddedb\\" isSection={true} showKey={true} />
</section>"
`;
